src/Chapter3.hs

```diff
@@ -1006,6 +1165,175 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Data Types
+
+-- Actions - all possible actions
+data Action = AttackFoe | CastSpell Int | DrinkHealthPotion Int | RunAway
+                deriving (Show)
+
+-- Test that an action is a knight action, to be used in smart constructor
+isKnightAction :: Action -> Bool
+isKnightAction AttackFoe             = True
+isKnightAction (CastSpell _)         = True
+isKnightAction (DrinkHealthPotion _) = True
+isKnightAction _                     = False
+
+-- Test that an action is a monster action, to be used in smart constructor
+isMonsterAction :: Action -> Bool
+isMonsterAction AttackFoe = True
+isMonsterAction RunAway   = True
+isMonsterAction _         = False
+
+-- Knight
+data Knight = Knight { knightName :: String
+                     , knightAttack :: Int
+                     , knightHealth :: Int
+                     , knightDefense ::Int
+                     , knightActions :: [Action]          
+                     } deriving (Show)
+
+-- smart constructor for Knight that enforces the correct subset of actions
```

Comment:
If you have two separate data types here, it makes sense to separate actions as well. then it would be safe to work with constructors itself without this smart constructor function 👍🏼